drivers/net/wireless/marvell/mwifiex/pcie_quirks.c

```diff
@@ -0,0 +1,248 @@
+		.matches = {
+			DMI_EXACT_MATCH(DMI_BIOS_VENDOR, "American Megatrends Inc."),
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "OEMB"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "OEMB"),
+		},
+		.driver_data = (void *)QUIRK_FW_RST_WSID_S3,
+	},
+	{
+		.ident = "Surface Pro 3",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 3"),
+		},
+		.driver_data = (void *)QUIRK_FW_RST_WSID_SP3,
+	},
+	{}
+};
+
+void mwifiex_initialize_quirks(struct pcie_service_card *card)
+{
+	const struct dmi_system_id *dmi_id;
+
+	dmi_id = dmi_first_match(mwifiex_quirk_table);
+	if (dmi_id)
+		card->quirks = (uintptr_t)dmi_id->driver_data;
+
+	if (card->quirks & QUIRK_FW_RST_D3COLD)
+		mwifiex_dbg(card->adapter, MSG,
+			    "quirk reset_d3cold enabled\n");
+	if (card->quirks & QUIRK_FW_RST_WSID_S3)
+		mwifiex_dbg(card->adapter, MSG,
+			    "quirk reset_wsid for Surface 3 enabled\n");
+	if (card->quirks & QUIRK_FW_RST_WSID_SP3)
+		mwifiex_dbg(card->adapter, MSG,
+			    "quirk reset_wsid for Surface Pro 3 enabled\n");
+}
+
+static void mwifiex_pcie_set_power_d3cold(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+
+	mwifiex_dbg(adapter, MSG, "putting into D3cold...\n");
+
+	pci_save_state(pdev);
+	if (pci_is_enabled(pdev))
+		pci_disable_device(pdev);
+	pci_set_power_state(pdev, PCI_D3cold);
+}
+
+static int mwifiex_pcie_set_power_d0(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	int ret;
+
+	mwifiex_dbg(adapter, MSG, "putting into D0...\n");
+
+	pci_set_power_state(pdev, PCI_D0);
+	ret = pci_enable_device(pdev);
+	if (ret) {
+		dev_err(&pdev->dev, "pci_enable_device failed\n");
+		return ret;
+	}
+	pci_restore_state(pdev);
+
+	return 0;
+}
+
+int mwifiex_pcie_reset_d3cold_quirk(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	struct pci_dev *parent_pdev = pci_upstream_bridge(pdev);
+
+	/* Power-cycle (put into D3cold then D0) */
+	mwifiex_dbg(adapter, MSG,
+		    "Using reset_d3cold quirk to perform FW reset\n");
+
+	/* We need to perform power-cycle also for bridge of wifi because
+	 * on some devices (e.g. Surface Book 1), the OS for some reasons
+	 * can't know the real power state of the bridge.
+	 * When tried to power-cycle only wifi, the reset failed with the
+	 * following dmesg log:
+	 * "Cannot transition to power state D0 for parent in D3hot".
+	 */
+	mwifiex_pcie_set_power_d3cold(pdev);
+	mwifiex_pcie_set_power_d3cold(parent_pdev);
+
+	mwifiex_pcie_set_power_d0(parent_pdev);
+	mwifiex_pcie_set_power_d0(pdev);
+
+	return 0;
+}
+
+int mwifiex_pcie_reset_wsid_quirk(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	char *path = ACPI_WSID_PATH;
```

Comment:
This should probably be `const`. You could also drop this and use the macro directly. For printing it, you can use
```c
dev_err(&pdev->dev, "No ACPI handle for path "ACPI_WSID_PATH "\n" );
```

Reply:
Regarding the constifying, `acpi_get_handle()` expects it to be `expected ‘acpi_string’ {aka ‘char *’}`. So, I'll keep it non-const.
```
  CC [M]  drivers/net/wireless/marvell/mwifiex/pcie_quirks.o
drivers/net/wireless/marvell/mwifiex/pcie_quirks.c: In function ‘mwifiex_pcie_reset_wsid_quirk’:
drivers/net/wireless/marvell/mwifiex/pcie_quirks.c:197:33: warning: passing argument 2 of ‘acpi_get_handle’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  197 |  status = acpi_get_handle(NULL, path, &handle);
      |                                 ^~~~
In file included from ./include/acpi/acpi.h:31,
                 from ./include/linux/acpi.h:22,
                 from drivers/net/wireless/marvell/mwifiex/pcie_quirks.c:6:
./include/acpi/acpixf.h:527:23: note: expected ‘acpi_string’ {aka ‘char *’} but argument is of type ‘const char *’
  527 |           acpi_string pathname,
      |           ~~~~~~~~~~~~^~~~~~~~
./include/acpi/acpixf.h:61:2: note: in definition of macro ‘ACPI_EXTERNAL_RETURN_STATUS’
   61 |  prototype;
      |  ^~~~~~~~~
```

With regard to using macro directly, I didn't know I can use it directly 😲 It seems that
```c
dev_err(&pdev->dev, "No ACPI handle for path %s\n", ACPI_WSID_PATH );
```
also works.

That said, I can't find the usage like this in upstream tree. Do you have the example? I'll keep current way for now.

Reply:
Static string concatenation is, for example, often used in `pr_fmt` definitions (e.g. [here](https://elixir.bootlin.com/linux/v5.8.12/source/drivers/cpufreq/pmac32-cpufreq.c#L12), where you have `KBUILD_MODNAME` and `fmt` as static strings, and the `": "` as separator).

Basically all that does is concatenate two static strings. You could also write `char *str = "abc" "def"` which then would get compiled to `char *str = "abcdef"`. The difference between
```c
dev_err(&pdev->dev, "No ACPI handle for path " ACPI_WSID_PATH "\n");
```
and
```c
dev_err(&pdev->dev, "No ACPI handle for path %s\n", ACPI_WSID_PATH);
```

is that the first will actually concatenate the strings at compile time, whereas the second will insert the string dynamically at runtime.

Reply:
And right, the `acpi_get_handle` not taking a `const char *` is a bit annoying...

Reply:
Looks like the upstream handles this in a few ways:

Using macro directly and using concatenating on print:
- https://github.com/torvalds/linux/blob/master/drivers/hwmon/asus_atk0110.c#L1213-L1219
```c
	/* RVLT: read voltage */
	status = acpi_get_handle(data->atk_handle, METHOD_OLD_READ_VLT, &ret);
	if (ACPI_SUCCESS(status))
		data->rvlt_handle = ret;
	else
		dev_dbg(dev, "method " METHOD_OLD_READ_VLT " not found: %s\n",
				acpi_format_exception(status));
```

Casting the const variable to `(char *)`
- https://github.com/torvalds/linux/blob/master/drivers/platform/x86/wmi.c#L162
- https://github.com/torvalds/linux/blob/master/drivers/platform/x86/asus-laptop.c#L396-L410
- https://github.com/torvalds/linux/blob/master/drivers/platform/x86/eeepc-laptop.c#L247-L252
```c
	status = acpi_get_handle(eeepc->handle, (char *)method,
				 handle);
	if (status != AE_OK) {
		pr_warn("Error finding %s\n", method);
		return -ENODEV;
	}
```

Casting to `(acpi_string)`
- https://github.com/torvalds/linux/blob/master/drivers/platform/x86/lg-laptop.c#L97
- https://github.com/torvalds/linux/blob/master/drivers/net/wireless/ath/ath10k/mac.c#L9001
```c
	status = acpi_get_handle(root_handle, (acpi_string)WRD_METHOD, &handle);
```

Using macro directly and using format on print:
- https://github.com/torvalds/linux/blob/master/drivers/acpi/acpica/dbtest.c#L187-L194
```c
		status =
		    acpi_get_handle(NULL, ACPI_DB_READ_METHOD, &read_handle);
		if (ACPI_FAILURE(status)) {
			acpi_os_printf
			    ("Could not obtain handle for debug method %s\n",
			     ACPI_DB_READ_METHOD);
			return;
		}
```

The last example is from `acpica` directory. So, I'll use the style and drop the `path` variable.

Reply:
Addressed with the new rebase.